Doesn't work with async traits

When using async traits (from the async-trait crate) both automock and mock! fail to generate valid code.

```
use async_trait::async_trait;
use mockall::*;

#[async_trait]
#[automock]
trait Widget {
    async fn foo();
}
```

> error[E0261]: use of undeclared lifetime name `'async_trait`
>  --> src\main.rs:5:1
>   |
> 5 | #[automock]
>   | ^^^^^^^^^^^ undeclared lifetime
> 
> error[E0582]: binding for associated type `Output` references lifetime `'async_trait`, which does not appear in the trait input types
>  --> src\main.rs:5:1
>   |
> 5 | #[automock]
>   | ^^^^^^^^^^^
> 
> error: aborting due to 2 previous errors

Switching the order:

```
#[automock]
#[async_trait]
trait Widget {
    async fn foo();
}
```

> error[E0706]: trait fns cannot be declared `async`
>  --> src\main.rs:4:11
>   |
> 4 | #[automock]
>   |           ^
> 
> error[E0706]: trait fns cannot be declared `async`
>  --> src\main.rs:4:11
>   |
> 4 | #[automock]
>   |           ^
> 
> error[E0195]: lifetime parameters or bounds on method `foo` do not match the trait declaration
>  --> src\main.rs:4:1
>   |
> 4 | / #[automock]
> 5 | | #[async_trait]
>   | | --------------
>   | | |
>   | | lifetimes in impl do not match this method in trait
> 6 | | trait Widget {
> 7 | |     async fn foo();
>   | |________________^ lifetimes do not match method in trait
> 
> error[E0195]: lifetime parameters or bounds on method `foo` do not match the trait declaration
>  --> src\main.rs:4:1
>   |
> 4 | / #[automock]
> 5 | | #[async_trait]
>   | | --------------
>   | | |
>   | | lifetimes in impl do not match this method in trait
> 6 | | trait Widget {
> 7 | |     async fn foo();
>   | |________________^ lifetimes do not match method in trait
> 
> error: aborting due to 2 previous errors

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,9 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased] - ReleaseDate
 ### Added
+- Compatibility with the `#[async_trait]` macro.
+  ([#183](https://github.com/asomers/mockall/pull/183))
+
 - Better support for non-Send types:
   * Added `return_const_st` for returning non-`Send` constants, similar to
     `returning_st`.

--- a/mockall/Cargo.toml
+++ b/mockall/Cargo.toml
@@ -33,6 +33,7 @@ predicates-tree = "1.0"
 mockall_derive = { version = "= 0.7.2", path = "../mockall_derive" }
 
 [dev-dependencies]
+futures = "0.3"
 serde = "1.0"
 serde_derive = "1.0"
 serde_json = "1.0"

--- a/mockall/Cargo.toml
+++ b/mockall/Cargo.toml
@@ -33,6 +33,7 @@ predicates-tree = "1.0"
 mockall_derive = { version = "= 0.7.2", path = "../mockall_derive" }
 
 [dev-dependencies]
+async-trait = "0.1.38"
 futures = "0.3"
 serde = "1.0"
 serde_derive = "1.0"

--- a/mockall/src/lib.rs
+++ b/mockall/src/lib.rs
@@ -45,6 +45,7 @@
 //! * [`Static methods`](#static-methods)
 //! * [`Foreign functions`](#foreign-functions)
 //! * [`Modules`](#modules)
+//! * [`Async Traits`](#async-traits)
 //! * [`Crate features`](#crate-features)
 //! * [`Examples`](#examples)
 //!
@@ -996,6 +997,38 @@
 //!         ctx.expect()
 //!             .returning(|x| i64::from(x + 1));
 //!         assert_eq!(5, inner::bar(4));
+//!     }
+//! }
+//! # fn main() {}
+//! ```
+//!
+//! ## Async Traits
+//!
+//! Async traits aren't yet (as of 1.47.0) a part of the Rust language.  But
+//! they're available from the
+//! [`async_trait`](https://docs.rs/async-trait/0.1.38/async_trait/) crate.
+//! Mockall is compatible with this crate, with two important limitations:
+//!
+//! * The `#[automock]` attribute must appear _before_ the `#[async_trait]`
+//! attribute.
+//!
+//! * The `#[async_trait]` macro must be imported with its canonical name.
+//!
+//! ```
+//! # use async_trait::async_trait;
+//! # use mockall::*;
+//! // async_trait works with both #[automock]
+//! #[automock]
+//! #[async_trait]
+//! pub trait Foo {
+//!    async fn foo(&self) -> u32;
+//! }
+//! // and mock!
+//! mock! {
+//!     pub Bar {}
+//!     #[async_trait]
+//!     trait Foo {
+//!         async fn foo(&self) -> u32;
 //!     }
 //! }
 //! # fn main() {}

--- a/mockall/tests/automock_async_trait.rs
+++ b/mockall/tests/automock_async_trait.rs
@@ -1,0 +1,22 @@
+// vim: tw=80
+//! An async trait, for use with Futures
+#![deny(warnings)]
+
+use async_trait::async_trait;
+use futures::executor::block_on;
+use mockall::*;
+
+#[automock]
+#[async_trait]
+pub trait Foo {
+    async fn foo(&self) -> u32;
+}
+
+
+#[test]
+fn return_const() {
+    let mut mock = MockFoo::new();
+    mock.expect_foo()
+        .return_const(42u32);
+    assert_eq!(block_on(mock.foo()), 42);
+}

--- a/mockall/tests/mock_async_fn.rs
+++ b/mockall/tests/mock_async_fn.rs
@@ -1,0 +1,20 @@
+// vim: tw=80
+//! A struct with an async function
+#![deny(warnings)]
+
+use futures::executor::block_on;
+use mockall::*;
+
+mock! {
+    pub Foo {
+        async fn foo(&self) -> u32;
+    }
+}
+
+#[test]
+fn return_const() {
+    let mut mock = MockFoo::new();
+    mock.expect_foo()
+        .return_const(42u32);
+    assert_eq!(block_on(mock.foo()), 42);
+}

--- a/mockall/tests/mock_async_trait.rs
+++ b/mockall/tests/mock_async_trait.rs
@@ -1,0 +1,28 @@
+// vim: tw=80
+//! An async trait, for use with Futures
+#![deny(warnings)]
+
+use async_trait::async_trait;
+use futures::executor::block_on;
+use mockall::*;
+
+#[async_trait]
+pub trait Foo {
+    async fn foo(&self) -> u32;
+}
+
+mock! {
+    pub Bar { }
+    #[async_trait]
+    trait Foo {
+        async fn foo(&self) -> u32;
+    }
+}
+
+#[test]
+fn return_const() {
+    let mut mock = MockBar::new();
+    mock.expect_foo()
+        .return_const(42u32);
+    assert_eq!(block_on(mock.foo()), 42);
+}

--- a/mockall/tools/allgen.sh
+++ b/mockall/tools/allgen.sh
@@ -12,5 +12,5 @@ cd ${TOOLSDIR}/../tests
 #exit 0
 for t in `ls *.rs | sed 's/\.rs//'`; do
 	env MOCKALL_DEBUG=1 cargo +nightly check --all-features --test $t > ${ODIR}/$t.rs || break;
-	rustfmt ${ODIR}/$t.rs ;
+	rustfmt --edition 2018 ${ODIR}/$t.rs ;
 done

--- a/mockall/tools/gen.sh
+++ b/mockall/tools/gen.sh
@@ -11,5 +11,5 @@ if [ -z "$t" ]; then
 	exit 2
 fi
 
-env MOCKALL_DEBUG=1 cargo check --all-features --test $t > ${PP}/$t.rs; rustfmt ${PP}/$t.rs
+env MOCKALL_DEBUG=1 cargo check --all-features --test $t > ${PP}/$t.rs; rustfmt --edition 2018 ${PP}/$t.rs
 echo "diff stat: " `diff ${PP_OLD}/$t.rs ${PP}/$t.rs | wc`

--- a/mockall_derive/src/lib.rs
+++ b/mockall_derive/src/lib.rs
@@ -443,15 +443,34 @@ fn find_lifetimes(ty: &Type) -> HashSet<Lifetime> {
     }
 }
 
-fn format_attrs(attrs: &[syn::Attribute], include_docs: bool) -> TokenStream {
-    let mut out = TokenStream::new();
-    for attr in attrs {
-        let is_doc = attr.path.get_ident().map(|i| i == "doc").unwrap_or(false);
-        if !is_doc || include_docs {
-            attr.to_tokens(&mut out);
+
+struct AttrFormatter<'a>{
+    attrs: &'a [Attribute],
+    doc: bool
+}
+
+impl<'a> AttrFormatter<'a> {
+    fn new(attrs: &'a [Attribute]) -> AttrFormatter<'a> {
+        Self {
+            attrs,
+            doc: true
         }
     }
-    out
+
+    fn doc(&mut self, allowed: bool) -> &mut Self {
+        self.doc = allowed;
+        self
+    }
+
+    fn format(&mut self) -> Vec<Attribute> {
+        self.attrs.iter()
+            .cloned()
+            .filter(|attr| self.doc ||
+                attr.path.get_ident()
+                .map(|i| i != "doc")
+                .unwrap_or(false)
+            ).collect()
+    }
 }
 
 /// Determine if this Pat is any kind of `self` binding

--- a/mockall_derive/src/lib.rs
+++ b/mockall_derive/src/lib.rs
@@ -446,15 +446,22 @@ fn find_lifetimes(ty: &Type) -> HashSet<Lifetime> {
 
 struct AttrFormatter<'a>{
     attrs: &'a [Attribute],
-    doc: bool
+    async_trait: bool,
+    doc: bool,
 }
 
 impl<'a> AttrFormatter<'a> {
     fn new(attrs: &'a [Attribute]) -> AttrFormatter<'a> {
         Self {
             attrs,
+            async_trait: true,
             doc: true
         }
+    }
+
+    fn async_trait(&mut self, allowed: bool) -> &mut Self {
+        self.async_trait = allowed;
+        self
     }
 
     fn doc(&mut self, allowed: bool) -> &mut Self {
@@ -462,13 +469,21 @@ impl<'a> AttrFormatter<'a> {
         self
     }
 
+    // XXX This logic requires that attributes are imported with their
+    // standard names.
     fn format(&mut self) -> Vec<Attribute> {
         self.attrs.iter()
             .cloned()
-            .filter(|attr| self.doc ||
-                attr.path.get_ident()
-                .map(|i| i != "doc")
-                .unwrap_or(false)
+            .filter(|attr|
+                ( self.doc ||
+                    attr.path.get_ident()
+                    .map(|i| i != "doc")
+                    .unwrap_or(false)
+                ) && (self.async_trait ||
+                    attr.path.get_ident()
+                    .map(|i| i != "async_trait")
+                    .unwrap_or(false)
+                )
             ).collect()
     }
 }

--- a/mockall_derive/src/mock_function.rs
+++ b/mockall_derive/src/mock_function.rs
@@ -324,7 +324,7 @@ pub(crate) struct MockFunction {
     /// Types of the method arguments
     argty: Vec<Type>,
     /// any attributes on the original function, like #[inline]
-    attrs: Vec<Attribute>,
+    pub attrs: Vec<Attribute>,
     /// Expressions that should be used for Expectation::call's arguments
     call_exprs: Vec<TokenStream>,
     /// Generics used for the expectation call
@@ -382,7 +382,7 @@ impl MockFunction {
     // Supplying modname is an unfortunately hack.  Ideally MockFunction
     // wouldn't need to know that.
     pub fn call(&self, modname: Option<&Ident>) -> impl ToTokens {
-        let attrs = self.format_attrs(true);
+        let attrs = AttrFormatter::new(&self.attrs).format();
         let call_exprs = &self.call_exprs;
         let (_, tg, _) = if self.is_method_generic() || self.is_static() {
             &self.egenerics
@@ -418,7 +418,7 @@ impl MockFunction {
             let outer_mod_path = self.outer_mod_path(modname);
             quote!(
                 // Don't add a doc string.  The original is included in #attrs
-                #attrs
+                #(#attrs)*
                 #vis #sig {
                     {
                         let __mockall_guard = #outer_mod_path::EXPECTATIONS
@@ -437,7 +437,7 @@ impl MockFunction {
         } else {
             quote!(
                 // Don't add a doc string.  The original is included in #attrs
-                #attrs
+                #(#attrs)*
                 #vis #sig {
                     self.#substruct_obj #name.#call#tbf(#(#call_exprs,)*)
                     .expect(#no_match_msg)
@@ -449,11 +449,13 @@ impl MockFunction {
 
     /// Return this method's contribution to its parent's checkpoint method
     pub fn checkpoint(&self) -> impl ToTokens {
-        let attrs = self.format_attrs(false);
+        let attrs = AttrFormatter::new(&self.attrs)
+            .doc(false)
+            .format();
         let inner_mod_ident = self.inner_mod_ident();
         if self.is_static {
             quote!(
-                #attrs
+                #(#attrs)*
                 {
                     let __mockall_timeses = #inner_mod_ident::EXPECTATIONS.lock()
                         .unwrap()
@@ -462,9 +464,8 @@ impl MockFunction {
                 }
             )
         } else {
-            let attrs = self.format_attrs(false);
             let name = &self.name();
-            quote!(#attrs { self.#name.checkpoint(); })
+            quote!(#(#attrs)* { self.#name.checkpoint(); })
         }
     }
 
@@ -476,7 +477,9 @@ impl MockFunction {
     // Supplying modname is an unfortunately hack.  Ideally MockFunction
     // wouldn't need to know that.
     pub fn context_fn(&self, modname: Option<&Ident>) -> impl ToTokens {
-        let attrs = self.format_attrs(false);
+        let attrs = AttrFormatter::new(&self.attrs)
+            .doc(false)
+            .format();
         let context_docstr = format!("Create a [`Context`]({}{}/struct.Context.html) for mocking the `{}` method",
             modname.map(|m| format!("{}/", m)).unwrap_or_default(),
             self.inner_mod_ident(),
@@ -486,7 +489,7 @@ impl MockFunction {
         let outer_mod_path = self.outer_mod_path(modname);
         let v = &self.call_vis;
         quote!(
-            #attrs
+            #(#attrs)*
             #[doc = #context_docstr]
             #v fn #context_ident() -> #outer_mod_path::Context #tg
             {
@@ -503,7 +506,9 @@ impl MockFunction {
     // Supplying modname is an unfortunately hack.  Ideally MockFunction
     // wouldn't need to know that.
     pub fn expect(&self, modname: &Ident) -> impl ToTokens {
-        let attrs = self.format_attrs(false);
+        let attrs = AttrFormatter::new(&self.attrs)
+            .doc(false)
+            .format();
         let name = self.name();
         let expect_ident = format_ident!("expect_{}", &name);
         let expectation_obj = self.expectation_obj();
@@ -541,7 +546,7 @@ impl MockFunction {
         quote!(
             #must_use
             #[doc = #docstr]
-            #attrs
+            #(#attrs)*
             #vis fn #expect_ident #ig(&mut self)
                -> &mut #modname::#expectation_obj
                #wc
@@ -573,29 +578,20 @@ impl MockFunction {
 
     pub fn field_definition(&self, modname: Option<&Ident>) -> TokenStream {
         let name = self.name();
-        let attrs = self.format_attrs(false);
+        let attrs = AttrFormatter::new(&self.attrs)
+            .doc(false)
+            .format();
         let expectations_obj = &self.expectations_obj();
         if self.is_method_generic() {
-            quote!(#attrs #name: #modname::#expectations_obj)
+            quote!(#(#attrs)* #name: #modname::#expectations_obj)
         } else {
             // staticize any lifetimes.  This is necessary for methods that
             // return non-static types, because the Expectation itself must be
             // 'static.
             let segenerics = staticize(&self.egenerics);
             let (_, tg, _) = segenerics.split_for_impl();
-            quote!(#attrs #name: #modname::#expectations_obj #tg)
+            quote!(#(#attrs)* #name: #modname::#expectations_obj #tg)
         }
-    }
-
-    pub fn format_attrs(&self, include_docs: bool) -> impl ToTokens {
-        let mut out = TokenStream::new();
-        for attr in &self.attrs {
-            let is_doc = attr.path.get_ident().map(|i| i == "doc").unwrap_or(false);
-            if !is_doc || include_docs {
-                attr.to_tokens(&mut out);
-            }
-        }
-        out
     }
 
     /// Human-readable name of the mock function
@@ -658,7 +654,9 @@ impl MockFunction {
 
     /// Generate code for this function's private module
     pub fn priv_module(&self) -> impl ToTokens {
-        let attrs = self.format_attrs(false);
+        let attrs = AttrFormatter::new(&self.attrs)
+            .doc(false)
+            .format();
         let common = &Common{f: self};
         let context = &Context{f: self};
         let expectation: Box<dyn ToTokens> = if self.return_ref {
@@ -696,7 +694,7 @@ impl MockFunction {
             Box::new(StaticRfunc{f: self})
         };
         quote!(
-            #attrs
+            #(#attrs)*
             #[allow(missing_docs)]
             pub mod #inner_mod_ident {
                 use super::*;

--- a/mockall_derive/src/mock_item_struct.rs
+++ b/mockall_derive/src/mock_item_struct.rs
@@ -74,8 +74,10 @@ impl Methods {
             .filter(|meth| !meth.is_static())
             .map(|meth| {
                 let name = meth.name();
-                let attrs = meth.format_attrs(false);
-                quote!(#attrs #name: Default::default())
+                let attrs = AttrFormatter::new(&meth.attrs)
+                    .doc(false)
+                    .format();
+                quote!(#(#attrs)* #name: Default::default())
             }).collect::<Vec<_>>()
     }
 
@@ -310,7 +312,9 @@ impl MockItemTraitImpl {
 
 impl ToTokens for MockItemTraitImpl {
     fn to_tokens(&self, tokens: &mut TokenStream) {
-        let attrs_nodocs = format_attrs(&self.attrs, false);
+        let attrs = AttrFormatter::new(&self.attrs)
+            .doc(false)
+            .format();
         let struct_name = &self.name;
         let (ig, tg, wc) = self.generics.split_for_impl();
         let modname = &self.modname;
@@ -330,7 +334,7 @@ impl ToTokens for MockItemTraitImpl {
             #[allow(non_camel_case_types)]
             #[allow(non_snake_case)]
             #[allow(missing_docs)]
-            #attrs_nodocs
+            #(#attrs)*
             struct #struct_name #ig #wc
             {
                 #(#field_definitions),*

--- a/mockall_derive/src/mock_item_struct.rs
+++ b/mockall_derive/src/mock_item_struct.rs
@@ -187,7 +187,9 @@ impl From<MockableStruct> for MockItemStruct {
 
 impl ToTokens for MockItemStruct {
     fn to_tokens(&self, tokens: &mut TokenStream) {
-        let attrs = &self.attrs;
+        let attrs = AttrFormatter::new(&self.attrs)
+            .async_trait(false)
+            .format();
         let struct_name = &self.name;
         let (ig, tg, wc) = self.generics.split_for_impl();
         let modname = &self.modname;
@@ -313,6 +315,7 @@ impl MockItemTraitImpl {
 impl ToTokens for MockItemTraitImpl {
     fn to_tokens(&self, tokens: &mut TokenStream) {
         let attrs = AttrFormatter::new(&self.attrs)
+            .async_trait(false)
             .doc(false)
             .format();
         let struct_name = &self.name;

--- a/mockall_derive/src/mock_trait.rs
+++ b/mockall_derive/src/mock_trait.rs
@@ -84,6 +84,7 @@ impl MockTrait {
     // Supplying modname is an unfortunately hack.  Ideally MockTrait
     // wouldn't need to know that.
     pub fn trait_impl(&self, modname: &Ident) -> impl ToTokens {
+        let attrs = &self.attrs;
         let (ig, tg, wc) = self.struct_generics.split_for_impl();
         let (_, t_tg, _) = self.trait_generics.split_for_impl();
         let calls = self.methods.iter()
@@ -101,6 +102,7 @@ impl MockTrait {
         let structname = &self.structname;
         let types = &self.types;
         quote!(
+            #(#attrs)*
             impl #ig #name #t_tg for #structname #tg #wc {
                 #(#types)*
                 #(#calls)*

